### ISSUE TITLE
Provide example on how to use `handleServerError`

### ISSUE THREAD

**Issue by omerlh** [endpoints4s/endpoints4s#1178]
I've tried using [`handleServerError`](https://endpoints4s.github.io/api/endpoints4s/http4s/server/EndpointsWithCustomErrors.html) so I can handle errors like invalid body passed and return 400 instead of 500 but I can't figure how to use it, can you help with example? I had issues with the right generic type to set for `org.http4s.Request` and I couldn't figure what to pass there.

Thanks for this really cool library!

**Comment by julienrf** [comment]
Could you please elaborate on the type of errors you would like to handle? `handleServerError` is called by endpoints4s when _an exception occurs during the request processing_. To handle validation errors, you may want to use `handleClientsErrors` instead.

**Comment by omerlh** [comment]
Yes. If you define a endpoint that receive a JSON body and you pass empty
body it returns incorrect 500 instead of 400

**Comment by julienrf** [comment]
I wonder if this is a bug in the http4s-server interpreter. The logic for decoding a JSON body is here:

https://github.com/endpoints4s/endpoints4s/blob/1b9ef64e66a7304cf7f337969b8deefd182d7788/http4s/server/src/main/scala/endpoints4s/http4s/server/JsonEntities.scala#L77-L88

I am not familiar with http4s myself but the `rethrowT` looks suspicious. Do you mind sending me a PR with a failing test [here](https://github.com/endpoints4s/endpoints4s/blob/master/algebras/testkit/src/main/scala/endpoints4s/algebra/server/JsonEntitiesFromSchemasTestSuite.scala)?

**Comment by omerlh** [comment]
I am trying to reproduce it with tests, but so far no dice.
This curl request reproduce it:
```
curl --location --request POST 'http://localhost:8080/produce' \
--data-raw ''
```
I tried many things, including passing empty array as body but it is not reproducing. 

This is my code: https://github.com/osskit/dafka-producer/blob/main/src/main/scala/ports/ProduceEndpoint.scala 
I had to use `JsonEntitiesFromCodecs` because I wanted to accept as parameter any object.

**Comment by julienrf** [comment]
Thank you for providing those details! I’ve investigated a bit and noticed a difference between Blaze and Ember. Blaze returns a 4xx error response in case of media type mismatch, whereas Ember returns a 5xx response. This is why we did not manage to reproduce your issue in the test suite (it uses Blaze whereas in your code you use Ember).

I am not sure whether the error should be handled by endpoints4s (instead of letting the underlying HTTP server implementation handle it).

The other way to solve your issue is to do it with `handleServerError`, but unfortunately `handleServerError` is not called is that situation, which is a bug in the http4s-server interpreter, IMHO. That can be fixed by adding yet another `recoverWith` here:

https://github.com/endpoints4s/endpoints4s/blob/1b9ef64e66a7304cf7f337969b8deefd182d7788/http4s/server/src/main/scala/endpoints4s/http4s/server/Endpoints.scala#L154-L174

I think I should at least fix that thing where we omit calling `handleServerError`, but there is still the open question on what should endpoints4s do when the incoming request has an incompatible type. In #544, we decided to check that all the server interpreters return a 415 response status (and there is no constraint on the format fo the response entity). If we continue with this approach, we should add an Ember-based test suite and make sure the behavior is consistent with the behavior we get with Blaze.

**Comment by omerlh** [comment]
Thanks for spending the time looking it and reproducing it!

**Comment by omerlh** [comment]
Update: there is similar error when not passing content-type header correctly - it will return 500 if content type is not application/json, I guess it is the same problem. I'll check if the same thing happens with Blaze

**Comment by julienrf** [comment]
Hey @omerlh. I’ve worked on a modification of the http4s-server interpreter in #1184 that would catch any failure wrapped in the underlying `IO[Response]` and let `handleServerError` handle it. I guess it would solve your problem. However, it breaks the current behavior of the interpreter and I would like to hear from you and @harpocrates what you think is the most appropriate thing to do.

Previously, when decoding the request entity failed with a failed `IO`, the failure was managed by the underlying http4s server, which, in the case of blaze, returned a 415 response status in case of media type mismatch. Now, in the same situation if we don’t give the underlying http4s server a chance to see that failure and call `handleServerError` instead, we return a 500 response status: https://github.com/endpoints4s/endpoints4s/actions/runs/5326705613/jobs/9649090178?pr=1184#step:6:1211.

What do you think endpoints4s should do here? I remember [being reluctant](https://github.com/endpoints4s/endpoints4s/pull/544#discussion_r414020314) to letting the underlying HTTP server return their own response in case of such failure. I do believe in the case of a media type mismatch we (endpoints4s) should return a 400 response instead of hoping that the underlying server will return a 415 response.

**Comment by omerlh** [comment]
Actually for the second case (not passing content type headers) 415 is the right status code, where for the first case (empty body) 400 is the right one. I think that if your code return the right status code it's improvement as it allows us to switch between Ember / Blaze without worrying about this kind of problems.
